### ISSUE TITLE
Support for Angular2

### ISSUE THREAD

**Issue by cajoy** [strongloop/loopback-sdk-angular#188]
Is anyone working on SDK for Angular2?

**Comment by arimus** [comment]
Yeah, I'm also curious if Strongloop is working on building out support for this or waiting for the community to go down that road first.

**Comment by daslicht** [comment]
+1

**Comment by bajtos** [comment]
Hi @cajoy, thank you for raising this request. We don't have any immediate plans to support specifically Angular2.

I haven't looked at the differences related to data access between Angular1 and Angular2, could you perhaps point me to a short overview?

In general, I would like to move away from angular-specific client API like `ngResource`, because the subtle differences between Angular client API and LoopBack server API are confusing and adds support overhead. I'd like to see a client API that is as close to the LoopBack server API as possible. I see two ways how to get there - a) modify the code generator to emit a different API or b) improve our universal/isomorphic client to support different HTTP backends (see https://github.com/strongloop/loopback/issues/1211 and of course https://github.com/strongloop/loopback/issues/989).

> @daslicht +1  What about React ?

That's off topic, please open a new GH issue in https://github.com/strongloop/loopback to discuss.

**Comment by cajoy** [comment]
For those who might be interesting I've created base model for extending. There is just couple of endpoints implemented..

https://github.com/cajoy/falloutlab.com/blob/master/client/scripts/models/BaseResource.ts


and model example:
 

https://github.com/cajoy/falloutlab.com/blob/master/client/scripts/models/PerkModel.ts


usage as simple as

```ts
                perkResource.upsert(perkModel)
                    .then((record: PerkModel) => {
                        console.log(record);
                    });
```

**Comment by CWSpear** [comment]
Now that Angular 2 is in Beta, it is worth discussing this topic further.

Angular 2 got rid of the `$scope.$apply` and similar concepts. The regular vanilla "[isomorphic LoopBack](https://docs.strongloop.com/display/public/LB/LoopBack+in+the+client)" in theory would work.

So you may not need an Angular specific SDK... you can just build up your isomorphic LoopBack story and provide a better onboarding for that way of using it (perhaps a lighter weight bundle or at least specific instructions on how to get started with Angular 2/various frameworks). (As an aside, perhaps "universal LoopBack" (see https://medium.com/@mjackson/universal-javascript-4761051b7ae9#.ud7guco7k) would be a better name.)

**Comment by JonnyBGod** [comment]
I agree with @CWSpear. And would like to add that api requests should be placed outside controllers and directives and live on services.

> "There is nothing specifically Angular about services. Angular itself has no definition of a service. There is no ServiceBase class."

Because Angular2 services are plain simple classes the isomorphic Loopback can be used.

I would like to propose adopting ES2015 for an improved implementation. This would bring the code up to date and we can easily compile to ES5 if needed. Also we would end up with much closer code to Angular2, React, and any recent framework.

As an extra, I would very much like to have a generator for redux-loopback. Happy to help with it. Redux e framework agnostic and is getting a lot of traction. After using it with React, I am currently playing with ng2-redux and it play really well with Angular2.

The main problem with "Isomorphic Loopback" is that it looks way long before it is usable. It is way too big for web usage and lacks customization. I will probably be testing swagger-codegen to generate javascript clients based on loopback-swagger definition until we get a good Javascript Loopback SDK.

**Comment by CWSpear** [comment]
So I have been having a hard time getting "isomorphic LoopBack" working. I tried building the bundle with browserify via the [example](https://docs.strongloop.com/display/LB/Running+LoopBack+in+the+browser) (it was almost 3 MB!) and couldn't actually ever make it work.

The current `lb-ng` tool created a build from the config in https://github.com/strongloop/loopback-example-isomorphic that was about 92 KB (and the *vast* majority of that was `jsdoc`, which would go away in minification)... when we're talking about the web here, that's a *massive* size difference.

So yeah, pretty much what @JonnyBGod said: some sort of generator like the `lb-ng` tool, but that just makes framework agnostic builds of the models so we can `require` them (or whatever we want!) and be on our merry little way.

The current state of "LoopBack in the browser" isn't really viable for anything outside of Angular.

**Comment by jonathan-casarrubias** [comment]
It would be awesome if a framework agnostic sdk is built, and event better if it provides some sort of adapter to customise how request are being done, perhaps you want to use standard http, but maybe you want to use RxJS, so you could be able to create an adapter and use it instead of the HTTP library.

This concept is being used by the socket.io clustering feature, where they provide a standard adapter that you can extend and a official redis-adapter, but now there are adapters for mongo so.. an sdk like this could work within any modern framework, and still can be flexible enough to be customised in the way requests are being done.

My two cents

**Comment by JonnyBGod** [comment]
After looking a bit deeper into this subject my main thought is: Why reinventing the wheel?

For most of us, creating api servers we probably need to provide many different clients for our customers as well as our own frontend clients. So why not bjust build this on top of swagger-codegen?

There are templates for most of the languages out there and also found templates for angularjs and a new (work in progress) one for angular2.

So in my opinion what we need is a small tool to help easily generate all the clients we need based on swagger-codegen. Something like:

```
lb-gen angular2 ./sdk/angular2
lb-gen java ./sdk/java
```

We already have loopback-swagger for generating the specs and all.

**Comment by ilkkanisula** [comment]
+1

**Comment by tuhoojabotti** [comment]
:+1:

**Comment by JoshElias** [comment]
+1

**Comment by phra** [comment]
any infos on when this PR (https://github.com/strongloop/loopback-sdk-angular-cli/pull/38) will be merged?

**Comment by yanisIk** [comment]
+1

**Comment by bajtos** [comment]
> For most of us, creating api servers we probably need to provide many different clients for our customers as well as our own frontend clients. So why not just build this on top of swagger-codegen?

:+1: 

We have already started to look into ways how to utilize swagger-codegen for creating a Swift client. Along the way, we discovered few issues in loopback-swagger, some of them are being fixed right now.

I'd like to encourage everybody subscribed to this thread: give swagger-codegen a try, let us know here how it works for you. If there is anything we can improve in swagger-codegen to make the generated client better, then let's fill GH issues in that project and work together to make them happen.

**Comment by phra** [comment]
@bajtos in the meanwhile you can accept [THIS](https://github.com/strongloop/loopback-sdk-angular/pull/208) and [THIS](https://github.com/strongloop/loopback-sdk-angular-cli/pull/38) in order to gain beta support for angular2 and let people around the world to test this feature (and hopefully finalize that template) from an **angular2 branch on the official repository** instead of to work on "_random_" forks.

just my two cents.

**Comment by phra** [comment]
does somebody already look at this project? https://github.com/wcandillon/swagger-js-codegen
it states: " A Swagger Codegen for typescript, nodejs & angularjs"

**Comment by bajtos** [comment]
@phra sorry for the long delay, we are a bit behind on reviewing & landing community pull requests. I am aware of swagger-js-codegen, I think it has been recently merged into [swagger-codegen](https://github.com/swagger-api/swagger-codegen) which supports many different languages.

In the long term, we would like to move away from code generators specific to loopback and invest our efforts into a more generic swagger-based code generator, i.e. swagger-codegen. If there are any small improvements that can be made in loopback to make the code generated by swagger-js-codegen look better, then we are happy to land them.

**Comment by asemoon** [comment]
+1

**Comment by kororo** [comment]
+1

**Comment by jlister76** [comment]
+1

**Comment by cristianurbano** [comment]
+1000

**Comment by sergiolepore** [comment]
@cristianurbano lol

**Comment by k1ng440** [comment]
+1m

**Comment by jboothe** [comment]
👍

**Comment by Vaelor** [comment]
+1 would love to see progress on this front

**Comment by jonathan-casarrubias** [comment]
They are currently working on V3 so I don't think this will be a priority so you may need to use @mnvx fork for a while

**Comment by ottojimb** [comment]
For newbies as me, how I resolve this:

- Go to http://0.0.0.0:3000/explorer/swagger.json and copy the response text.
- Go to http://codebeautify.org/jsonviewer and paste, after push the beautify button and copy again the response text (this is for easy debug on the next step)
- Go to http://editor.swagger.io/#/ and click on file->import json. Paste the text. On this step the editor can show errors: only click "import" and it will show the errors by each line. I got errors with containers because the parameters are required (with true) and the generator was created this as "required: false". Example (lines 5983 and 5987): 
<img width="1275" alt="Screenshot 1" src="https://cloud.githubusercontent.com/assets/1256428/15868378/3855dc56-2cad-11e6-9848-9c214185c179.png">
- Another error with containers is "Path parameter is declared but is not defined: container" Example:
<img width="1276" alt="Screenshot 2" src="https://cloud.githubusercontent.com/assets/1256428/15868802/0489ccdc-2caf-11e6-8be0-b07d055095da.png">
To solve: fill the parameter values and drop the empty braces:
```
        - name: container
          in: path
          required: true
          type: string
```
Will be:
<img width="1271" alt="Screenshot 3" src="https://cloud.githubusercontent.com/assets/1256428/15868933/886fe162-2caf-11e6-87ab-a0a8af6b5078.png">
- Once you have fixed all errors, only click on "Generate client" and "Typescript Angular2". 

Advantages:

- The generated files are well documented and returning observables (no more time wasted writing the client libraries: swagger support many of them)
- The generated files separate the basic model files from api files
- The documentation is ready for the backend app. (no more time wasted)

<img width="492" alt="Screenshot 4" src="https://cloud.githubusercontent.com/assets/1256428/15869153/b9bc57a4-2cb0-11e6-9a5c-4e031feccbde.png">


I hope this can help to somebody. Sorry for my bad english.

**Comment by mi9rom** [comment]
Hi @ottojimb ,
Unfortunately you need to port the LoopBackAuth as well + custom login,logout, unless you don't care included functionalities.
I'd been looking for sth like this but failed. 
Regards :)

**Comment by jonathan-casarrubias** [comment]
@mi9rom I have the same concern, there is logic behind to really integrate with LoopBack.. I'm still not sure using swagger is the right option

**Comment by JonnyBGod** [comment]
The loopback specific logic can be dealt with separately.
We could go with many different approaches, create a small generators for that if this we want to try and tackle custom configs or just create a repos with templates.

I would probably go with the second option and create a tutorial on how to extend these templates with custom options. This part of the projects tend not to change that much. After you create the base behaviour of your apis the auth process his pretty static. Changes are model related. This is my experience at least. Correct me if I am wrong.

**Comment by jonathan-casarrubias** [comment]
I still disagree, the same Angular2 SDK will have differences in terms of implementation between NativeScript2, Angular2 For Web or Angular Universal.. I honestly see too much work in the second option and too long to wait a release, we have been using qeti and mnvx's fork for a while now and is the fastest way  for those -like me- who can not have such a slow rhythm.

I extended from their fork and working on my necessities that goes beyond auth, it includes socket io communication for the [loopback-component-pubsub](https://www.npmjs.com/package/loopback-component-pubsub), and much more.. but anyway, I think is time to have a community based builder rather than wait forever before we can even get just 1 Angular 2 compatible skd for web or spend time getting small pieces together.

Anyway, that is just my opinion.

Cheers,
Jon

**Comment by jonathan-casarrubias** [comment]
I just published a tutorial of how to use these modules, for those who may be interested:

[The Ultimate Guide for building RTA](http://mean.expert/2016/06/09/angular-2-ultimate-real-time/)

Cheers!!
Jon.

**Comment by ottojimb** [comment]
@mi9rom, @jonathan-casarrubias, thank you!! I'm agree with another real SDK for angular2. Completely unaware of other options and my intention was sharing how I was doing.

**Comment by jonathan-casarrubias** [comment]
@ottojimb thanks for your support.

Currently the most mature version of the builder is for NativeScript, please beware that I just gave a small amount of support for the Angular2 for Web Version.. I think it works, but is not optimal... Since I think I'm done this week with the NativeScript one, I will work next week on porting everything so we have same functionality according the Web Environment, for this I mean... to support web sockets and loopback configuration, that was not implemented by qeti and mnvx.

Still I think the Angular 2 SDK for Web works, so if you want to start playing with it is already compatible with Angular 2 RC.

I will be announcing when the Web Version is fully ported...

Cheers,
Jon!

**Comment by bajtos** [comment]
For anybody stumbling across this discussion while looking for an Angular2 client for LoopBack: there are several community-driven solutions available:
 - [loopback-sdk-builder](https://www.npmjs.com/package/loopback-sdk-builder)
 - Use Swagger Editor to generate a client, see https://www.npmjs.com/package/loopback-sdk-builder

**Comment by djabif** [comment]
Hi @jonathan-casarrubias thanks for sharing your solution!
Will it work with Angular2 for TypeScript?

**Comment by jonathan-casarrubias** [comment]
Hey @djabif yes!!! If you build an SDK for Angular2 it will be fully compatible with TypeScript.

Cheers!
Jon

**Comment by levantoanbkit** [comment]
Dear sir @jonathan-casarrubias , @bajtos 
I am working on Ionic2 Beta version and also want to integrate with Loopback framework. I am very interested when was known about loopback-sdk-builder, but I am not sure whether can it work with Ionic2 Beta (AngularJS 2)? Please help me, can you share your experiences about my difficulties?
Thank you so much!

**Comment by jonathan-casarrubias** [comment]
@levantoanbkit it should work with ionic by using the ng4web driver. Please read the loopback-sdk-builder README and build your SDK for Ionic as if it were for web.

Cheers!
Jon

**Comment by levantoanbkit** [comment]
Thank you so much! @jonathan-casarrubias

**Comment by jonathan-casarrubias** [comment]
@bajtos @davidcheung I'm struggling with a current validation from loopback/strong-remoting to generate model definitions for those models who are private, we do not want to expose the services but in order to support a fully typed sdk, we need the model definitions.

Can you take a look into this issue? I know how could be solved but it need a small core modification of the framework

https://github.com/jonathan-casarrubias/loopback-sdk-builder/issues/47

Cheers
Jon

**Comment by jonathan-casarrubias** [comment]
![LoopBack SDK Builder](https://storage.googleapis.com/mean-expert-images/sdk-builder.jpg)

Aloha LoopBack SDK Builder 🚀 
===========

I'm really exited to notify anyone interested in generating a [LoopBack](http://loopback.io) SDK for [Angular 2](http://angular.io) that after several months of hard work the first [LoopBack SDK Builder](https://www.npmjs.com/package/@mean-expert/loopback-sdk-builder) Stable Release is already online.

Featuring [Angular 2](http://angular.io) Final API with [NgModule](https://angular.io/docs/ts/latest/guide/ngmodule.html), New [Wiki Docs](https://github.com/mean-expert-official/loopback-sdk-builder/wiki), new mascot and much more:

# Features

- Support for Angular 2 (Final API w/ NgModules & Backwards compatibility < RC5).
- Support for TypeScript (Fully Typed).
- Built in Interfaces and Models.
- Extendable Models for custom logic.
- Enables Support for Real-Time Applications [loopback-component-pubsub](https://www.npmjs.com/package/loopback-component-pubsub)
- Built in LoopBack Authentication.
- Built in Support for LoopBack Query Language [Querying Data](https://docs.strongloop.com/display/public/LB/Querying+data)
- Built in API Servics.
- Built in Platform Specific Drivers (Angular2 for web, NativeScript2, [~~Angular Universal~~](https://github.com/mean-expert-official/loopback-sdk-builder/issues/36)).
- Built in CLI Tool for builder.
- Built in Logger Service.
- Blacklist mechanism to select which models or methods generate.
- Ability to point models to different url domains (not global baseUrl)
- IO Heartbeating to avoid disconnections.
- Small foot print 100k per generated SDK (Will increase depending on number of models).

**Comment by AurelieV** [comment]
@jonathan-casarrubias great news !
Does the SDK support cookie authentication or does it need token as the angular1 version does?

**Comment by jonathan-casarrubias** [comment]
@AurelieV for now it does use token authentication, but I'm thinking in support cookie authentication when [Angular Universal](http://universal.angular.io) is supported by the builder, that is in the roadmap for version 2.1.0, check [here](https://github.com/mean-expert-official/loopback-sdk-builder/issues/36)

**Comment by phra** [comment]
big up for @jonathan-casarrubias!

**Comment by AurelieV** [comment]
@jonathan-casarrubias be able to choose will be very appreciate :). I will check if it's easy or not (only have to not send the token so)

**Comment by jonathan-casarrubias** [comment]
@AurelieV is a challenge, because [Angular Universal](http://universal.angular.io) only renders the page the first time is loaded in the browser, then [Angular 2](http://angular.io) takes control in the front end, so it can be partially implemented. Please follow the [thread](https://github.com/mean-expert-official/loopback-sdk-builder/issues/36) for progress on this topic.

Cheers!!!
Jon

**Comment by levantoanbkit** [comment]
Thank your team so much!

**Comment by bajtos** [comment]
> I'm really exited to notify anyone interested in generating a [LoopBack](http://loopback.io) SDK for [Angular 2](http://angular.io); that after several months of hard work the first [LoopBack SDK Builder](https://www.npmjs.com/package/@mean-expert/loopback-sdk-builder) Stable Release is already online.

> Featuring [Angular 2](http://angular.io) Final API with [NgModule](https://angular.io/docs/ts/latest/guide/ngmodule.html) Support, New [Wiki Docs](https://github.com/mean-expert-official/loopback-sdk-builder/wiki), New Mascot and More:

@jonathan-casarrubias congratulations! I am personally very happy that you managed to make this happen and wish you many happy users & contributors.

@crandmck I am proposing to add a link to our documentation pointing to this Angular2 SDK, what do you think?

I am closing this particular github issue as done.

**Comment by jonathan-casarrubias** [comment]
Hey @bajtos thanks a lot for your nice words!!!

Also.. It would be wonderful if you guys can add a link in the official documentation for the Angular 2 SDK.

There is a good amount of people migrating already to NG2 and would provide a better experience for them to land directly from the official docs and by not digging in the web.

Again thanks for your words and your support.

Cheers!!!
Jon

**Comment by crandmck** [comment]
Added a doc page: https://docs.strongloop.com/display/LB/Angular+2+SDK

NOTE: We're in the process of migrating the docs to GitHub in https://github.com/strongloop/loopback.io and you'll be able to add more details there, if you wish, once we start accepting PRs.  (Still lots of work to do on http://loopback.io/doc/ so not yet!).

**Comment by jonathan-casarrubias** [comment]
Hey @crandmck thanks for adding the section, and absolutely I can contribute to the documentation once you start accepting pull request, I will be looking forward for it

Cheers!
Jon
